server/etcdserver/server.go

```diff
@@ -207,8 +207,10 @@ type EtcdServer struct {
 	term              uint64 // must use atomic operations to access; keep 64-bit aligned.
 	lead              uint64 // must use atomic operations to access; keep 64-bit aligned.
 
-	consistIndex cindex.ConsistentIndexer // consistIndex is used to get/set/save consistentIndex
-	r            raftNode                 // uses 64-bit atomics; keep 64-bit aligned.
+	consistentIdx  uint64                   // must use atomic operations to access; keep 64-bit aligned.
```

Comment:
Why we are adding this instead of use internals of 'consistIndex' ?

Reply:
I just added `consistentIdx` and `consistentTerm` to temporarily save the e.Index and e.Term, and in the txPostLockHook they are saved to the ConsistentIndex and Term by calling `s.consistIndex.SetConsistentIndex(cIdx, term)`. The original `consistIndex` keeps unchanged. 

```

	consistentIdx  uint64                   // must use atomic operations to access; keep 64-bit aligned.
	consistentTerm uint64                   // must use atomic operations to access; keep 64-bit aligned.
	consistIndex   cindex.ConsistentIndexer // consistIndex is used to get/set/save consistentIndex
```

Reply:
I still don't understand.  cindex.ConstentIndexer has 2 purposes: 
  a) storing the cindex temporarilly `(consistentIndex, term)` variables. 
  b) Flushing it to the backend with 'UnsafeSave' is called. 

Why the hook cannot use UnsafeSave to flush it and  `consistIndex.SetConsistentIndex()` as the temporary storage ?

Reply:
Previously, we save the consistent_index and term into cindex by calling s.consistIndex.SetConsistentIndex(e.Index, e.Term) in applyEntries, and persists the values into db in preCommitHook. So the data path is:
```
e.Index, e.Term --> ci.consistentIndex, ci.term  --> db
```

Now, I added one more layer of cache, because we can't save the `e.Index and e.Term` into `ci.consistentIndex and ci.term ` directly in applyEntries, 
```
e.Index, e.Term --> s.consistentIdx, s.consistentTerm -->  ci.consistentIndex, ci.term --> db
```

Reply:
> we can't save the e.Index and e.Term into ci.consistentIndex and ci.term 

Solution I'm thinking about is following: 
 1. We remove the call from the hook: https://github.com/etcd-io/etcd/blob/d24ef3ac20ebfb7fa1f7c2c1a693c5c80ffcb136/server/storage/hooks.go#L44-L45
2. We call it directly from (SetTxPostLockHook)

I might be missing something obvious, but I think that: 
 1) We are getting simpler code without another `s.consistentIdx` field exposed that one can use as authoritative state of the raft. The state exposed in EtcdServer and not encapsulated in some object/interface with clear purpose is likely to be misused as the class is widely provided within the huge package. 
 2) We might loose some performance, because we will be setting the `key` in a single bolt transaction multiple times. I don't think it's a significant performance gain, as (I think) bolt in scope of a single transaction is working in-place on the same page. 

I think that 1) is more important than 2).  Is there another  motivation I'm missing ?
If there is ... we should at least pack the SetTxPostLockHook and the fields in a single object that encapsulates the logic.

Reply:
Yes, we are on the same page now. Just as I mentioned in the third proposal in [discussion_r840127417](https://github.com/etcd-io/etcd/pull/13854#discussion_r840127417), it doesn't make much sense to persist the consistent_index in each apply (not the bbolt db transaction), because it will cause decrease of performance. If we follow this way, then what's value of `OnPreCommitUnsafe`?

I agree that introducing s.consistentIdx into Etcdserver is not a good way of encapsulation.  Probably we can move the two fields into [consistentIndex](https://github.com/etcd-io/etcd/blob/d24ef3ac20ebfb7fa1f7c2c1a693c5c80ffcb136/server/etcdserver/cindex/cindex.go#L47). WDYT?

Reply:
consistentIndex or new object seems good to me. 
 we can call them: 
     "applyingIndex" and "applyingTerm" for example to distingush from "appliedIndex & appliedTerm". 
     
 I think that at the time the code was written... it seemed as not difference/cheap optimisation to delay the write to "on commit". There was no measurement whether it's actual improvement.

Reply:
Yes, after second thought I'd agree that there shouldn't much performance gain by introducing the `OnPreCommitUnsafe`, because the bbolt DB loads a page to a node only once in each transaction, and any following updates are only in the node (in memory).  The only gain is reducing multiple operations on node to only one. But I agree we need to run some benchmark to compare the performance difference. 

I will take a look at this once this issue is resolved. Afterwards, we can make a decision whether or not we should remove the `OnPreCommitUnsafe`.

But for now, let's just keep it as it's. 

`"applyingIndex" and "applyingTerm"` make perfect sense to me. Will move them into `consistentIndex`. Thanks.

Reply:
Thanks & Done.

Reply:
Please take a look at the last commit in this PR.